### PR TITLE
Fix bug in projectAsLibrary test

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
- Fix bug in projectAsLibrary test

### How has this been tested?
- Existing test now completes in 1 seconds instead of getting stuck. Not sure how it ever passes in CI, but there it was taking a huge time to complete.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/DooFile.cs [hunk 1]
```diff
@@ -168,18 +169,21 @@ public class ManifestData {
...
   public static DafnyOptions? CheckAndGetLibraryOptions(ErrorReporter reporter,
     Uri libraryFile,
     DafnyOptions options, IToken origin,
     IDictionary<string, object> libraryOptions,
-    Dictionary<Option, OptionCompatibility.OptionCheck> additionalOptions) {
+    ImmutableDictionary<Option, OptionCompatibility.OptionCheck> additionalOptions) {
     var result = new DafnyOptions(options);
     var success = true;
+    var totalOptions = additionalOptions.Merge(OptionChecks,
+      (_, _) => throw new InvalidOperationException());
```
Comment (robin-aws) [inline Source/DafnyCore/DooFile.cs:183]: For next opportunity: having a MergeDistinct variant would make this more readable (I had to go look up what the function value was for to understand)

**Review by robin-aws — COMMENTED** [review]
Fix itself LGTM - checking my understanding, the problem was incorrectly thinking a dependency did not have compatible options, and therefore falling back to verifying the dependency when that wasn't the intent?

I'm a little worried this is a big foot gun for users to hit in the same way if so.

**Comment by keyboardDrummer** [comment]
> Fix itself LGTM - checking my understanding, the problem was incorrectly thinking a dependency did not have compatible options, and therefore falling back to verifying the dependency when that wasn't the intent?
> 

No. The problem was dat `--libraries` was considered a non-semantic option, meaning it is copied from the root project to the dependency, causing the dependency to have itself as a dependency which caused a cycle in the compilation process.

I've refactored this PR, so the resulting code is now simpler.

**Review by robin-aws — DISMISSED** [review]
I do think we need another category of options rather than special-casing `--library`, but this is a good improvement for now.

I'm also very worried about a test that apparently should fail but instead takes a long time and then passes! But I don't think leaving this test in this state is a good idea just to not lose track of that problem.

**Review by ssomayyajula — APPROVED** [review]

**Pull request merged**